test_install/test_install.cpp

```diff
@@ -42,13 +44,128 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+    auto x_clone = gko::clone(x);
+    test->read(A_raw);
+    test->apply(b, gko::lend(x_clone));
+    // x_clone has the device result if using HIP or CUDA, otherwise it is
+    // reference only
+
+#if defined(HAS_HIP) || defined(HAS_CUDA)
+    // If we are on a device, we need to run a reference test to compare against
+    auto exec_ref = exec->get_master();
+    auto test_ref = Mtx::create(exec_ref);
+    auto x_ref = gko::clone(exec_ref, x);
+    test_ref->read(A_raw);
+    test_ref->apply(b, gko::lend(x_ref));
+
+    // Actually check that `x_clone` is similar to `x_ref`
+    auto x_clone_ref = gko::clone(exec_ref, gko::lend(x_clone));
+    assert_similar_matrices(gko::lend(x_clone_ref), gko::lend(x_ref), 1e-14);
+#endif  // defined(HAS_HIP) || defined(HAS_CUDA)
+#endif  // HAS_REFERENCE
+}
+
+
+template <typename Solver>
+void check_solver(std::shared_ptr<gko::Executor> exec,
+                  gko::matrix_data<double> &A_raw, gko::matrix::Dense<> *b,
+                  gko::matrix::Dense<> *x)
+{
+    using Mtx = gko::matrix::Csr<>;
+    auto A =
+        gko::share(Mtx::create(exec, std::make_shared<Mtx::load_balance>()));
+
+    auto num_iters = 20u;
+    double reduction_factor = 1e-7;
+    auto solver_gen =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(num_iters).on(
+                    exec),
+                gko::stop::ResidualNorm<>::build()
+                    .with_reduction_factor(reduction_factor)
+                    .on(exec))
+            .on(exec);
+#if HAS_REFERENCE
+    A->read(A_raw);
+    auto x_clone = gko::clone(x);
+    solver_gen->generate(A)->apply(b, gko::lend(x_clone));
+    // x_clone has the device result if using HIP or CUDA, otherwise it is
+    // reference only
+
+#if defined(HAS_HIP) || defined(HAS_CUDA)
+    // If we are on a device, we need to run a reference test to compare against
+    auto exec_ref = exec->get_master();
+    auto A_ref =
+        gko::share(Mtx::create(exec_ref, std::make_shared<Mtx::load_balance>()));
+    A_ref->read(A_raw);
+    auto solver_gen_ref =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(num_iters).on(
+                    exec_ref),
+                gko::stop::ResidualNorm<>::build()
+                    .with_reduction_factor(reduction_factor)
+                    .on(exec_ref))
+            .on(exec_ref);
+    auto x_ref = gko::clone(exec_ref, x);
+    solver_gen->generate(A_ref)->apply(b, gko::lend(x_ref));
+
+    // Actually check that `x_clone` is similar to `x_ref`
+    auto x_clone_ref = gko::clone(exec_ref, gko::lend(x_clone));
+    assert_similar_matrices(gko::lend(x_clone_ref), gko::lend(x_ref), 1e-12);
+#endif  // defined(HAS_HIP) || defined(HAS_CUDA)
+#endif  // HAS_REFERENCE
+}
+
+
 // core/base/polymorphic_object.hpp
 class PolymorphicObjectTest : public gko::PolymorphicObject {};
 
 
 int main(int, char **)
 {
```

Comment:
nit: Since we don't use the parameter, can we simply change it to:
```suggestion
int main()
{
```